File: backend/validator.py
"""SpatioTrust validator — pure stdlib (no Open3D/PyTorch).

Mirrors src/lib/validator.ts byte-for-byte in spirit:
  1) AABB + centroid
  2) Base set = lowest 10% of height
  3) Centroid (x,z) must lie inside base footprint AND base mass >= 12%
  4) Floating-mass scan over 10 y-slices (>8% mass with 0 directly below = anomaly)
  5) Weighted confidence + zk_mock_hash = sha256(canonical_json + status)
"""
import hashlib, json, math, time

def _round(v: float) -> float:
    return round(v, 3)

def validate_point_cloud(points):
    if not points:
        raise ValueError("Empty point cloud")

    n = len(points)
    x_min = y_min = z_min = math.inf
    x_max = y_max = z_max = -math.inf
    cx = cy = cz = 0.0
    for p in points:
        x, y, z = p["x"], p["y"], p["z"]
        x_min = min(x_min, x); x_max = max(x_max, x)
        y_min = min(y_min, y); y_max = max(y_max, y)
        z_min = min(z_min, z); z_max = max(z_max, z)
        cx += x; cy += y; cz += z
    cx /= n; cy /= n; cz /= n
    y_range = max(y_max - y_min, 1e-6)

    base_thresh = y_min + 0.10 * y_range
    bx_min = bz_min = math.inf
    bx_max = bz_max = -math.inf
    bcx = bcz = 0.0
    base_idx = []
    for i, p in enumerate(points):
        if p["y"] <= base_thresh:
            base_idx.append(i)
            bx_min = min(bx_min, p["x"]); bx_max = max(bx_max, p["x"])
            bz_min = min(bz_min, p["z"]); bz_max = max(bz_max, p["z"])
            bcx += p["x"]; bcz += p["z"]
    base_count = len(base_idx)
    base_ratio = base_count / n
    if base_count:
        bcx /= base_count; bcz /= base_count

    slack_x = 0.05 * (x_max - x_min)
    slack_z = 0.05 * (z_max - z_min)
    centroid_inside = (
        base_count > 0
        and bx_min - slack_x <= cx <= bx_max + slack_x
        and bz_min - slack_z <= cz <= bz_max + slack_z
    )
    enough_base_mass = base_ratio >= 0.12
    centroid_supported = centroid_inside and enough_base_mass

    slice_count = 10
    slice_h = y_range / slice_count
    counts = [0] * slice_count
    slice_pts = [[] for _ in range(slice_count)]
    for i, p in enumerate(points):
        s = int((p["y"] - y_min) / slice_h)
        s = min(max(s, 0), slice_count - 1)
        counts[s] += 1
        slice_pts[s].append(i)

    floating_mass = False
    anomaly = []
    for s in range(2, slice_count):
        if counts[s] / n > 0.08 and counts[s - 1] == 0:
            floating_mass = True
            anomaly.extend(slice_pts[s])

    support_score = 1.0 if centroid_supported else 0.2
    float_score = 0.1 if floating_mass else 1.0
    mass_score = min(1.0, base_ratio / 0.20)
    confidence = max(0.0, min(1.0, 0.45 * support_score + 0.40 * float_score + 0.15 * mass_score))
    status = "pass" if confidence >= 0.7 and not floating_mass and centroid_supported else "fail"
    anomaly_detected = floating_mass or not centroid_supported

    sorted_pts = sorted(points, key=lambda p: (p["x"], p["y"], p["z"]))
    payload = json.dumps(sorted_pts, separators=(",", ":")) + ":" + status
    digest = hashlib.sha256(payload.encode("utf-8")).hexdigest()

    return {
        "status": status,
        "confidence": round(confidence, 3),
        "anomaly_detected": anomaly_detected,
        "zk_mock_hash": "0x" + digest,
        "metrics": {
            "total": n,
            "baseCount": base_count,
            "baseRatio": round(base_ratio, 3),
            "centroid": {"x": _round(cx), "y": _round(cy), "z": _round(cz)},
            "baseCentroid": {"x": _round(bcx), "z": _round(bcz)},
            "centroidSupported": centroid_supported,
            "floatingMass": floating_mass,
            "anomalyIndices": anomaly,
            "bounds": {
                "min": {"x": _round(x_min), "y": _round(y_min), "z": _round(z_min)},
                "max": {"x": _round(x_max), "y": _round(y_max), "z": _round(z_max)},
            },
        },
    }

if __name__ == "__main__":
    import sys
    data = json.load(open(sys.argv[1]))
    pts = data if isinstance(data, list) else data["points"]
    t0 = time.time()
    print(json.dumps(validate_point_cloud(pts), indent=2))
    print(f"# elapsed: {(time.time()-t0)*1000:.1f} ms", file=sys.stderr)

File: src/lib/validator.ts
import { sha256Hex } from "./hash";

export type Point = { x: number; y: number; z: number };

export type ValidationMetrics = {
  total: number;
  baseCount: number;
  baseRatio: number;
  centroid: Point;
  baseCentroid: { x: number; z: number };
  centroidSupported: boolean;
  floatingMass: boolean;
  anomalyIndices: number[];
  bounds: { min: Point; max: Point };
};

export type ValidationResult = {
  status: "pass" | "fail";
  confidence: number;
  anomaly_detected: boolean;
  zk_mock_hash: string;
  metrics: ValidationMetrics;
};

export type ValidatorOpts = {
  /** 0..0.5 — extra slack when checking centroid lies inside base footprint. */
  baseSupportTolerance?: number;
  /** 0..1 — pass/fail confidence threshold (higher = stricter). */
  confidenceSensitivity?: number;
};

export async function validatePointCloud(
  points: Point[],
  opts: ValidatorOpts = {}
): Promise<ValidationResult> {
  const slackPct = clamp01(opts.baseSupportTolerance ?? 0.15) * 1; // already 0..0.5
  const sensitivity = clamp01(opts.confidenceSensitivity ?? 0.7);
  if (!Array.isArray(points) || points.length === 0) {
    throw new Error("Empty point cloud");
  }

  // 1. AABB + centroid
  let xMin = Infinity, yMin = Infinity, zMin = Infinity;
  let xMax = -Infinity, yMax = -Infinity, zMax = -Infinity;
  let cx = 0, cy = 0, cz = 0;
  for (const p of points) {
    if (p.x < xMin) xMin = p.x; if (p.x > xMax) xMax = p.x;
    if (p.y < yMin) yMin = p.y; if (p.y > yMax) yMax = p.y;
    if (p.z < zMin) zMin = p.z; if (p.z > zMax) zMax = p.z;
    cx += p.x; cy += p.y; cz += p.z;
  }
  const n = points.length;
  cx /= n; cy /= n; cz /= n;

  const yRange = Math.max(yMax - yMin, 1e-6);

  // 2. Base set: lowest 10% of height
  const baseThreshold = yMin + 0.10 * yRange;
  const baseIndices: number[] = [];
  let bxMin = Infinity, bzMin = Infinity, bxMax = -Infinity, bzMax = -Infinity;
  let bcx = 0, bcz = 0;
  for (let i = 0; i < n; i++) {
    const p = points[i];
    if (p.y <= baseThreshold) {
      baseIndices.push(i);
      if (p.x < bxMin) bxMin = p.x; if (p.x > bxMax) bxMax = p.x;
      if (p.z < bzMin) bzMin = p.z; if (p.z > bzMax) bzMax = p.z;
      bcx += p.x; bcz += p.z;
    }
  }
  const baseCount = baseIndices.length;
  const baseRatio = baseCount / n;
  if (baseCount > 0) { bcx /= baseCount; bcz /= baseCount; }

  // 3. Centroid x,z must lie inside base footprint (with small slack)
  const slackX = slackPct * (xMax - xMin || 1);
  const slackZ = slackPct * (zMax - zMin || 1);
  const centroidInside =
    baseCount > 0 &&
    cx >= bxMin - slackX && cx <= bxMax + slackX &&
    cz >= bzMin - slackZ && cz <= bzMax + slackZ;
  const enoughBaseMass = baseRatio >= 0.12;
  const centroidSupported = centroidInside && enoughBaseMass;

  // 4. Floating-mass scan: 10 slices above base
  const sliceCount = 10;
  const sliceHeight = yRange / sliceCount;
  const sliceCounts = new Array(sliceCount).fill(0);
  const slicePoints: number[][] = Array.from({ length: sliceCount }, () => []);
  for (let i = 0; i < n; i++) {
    const p = points[i];
    let s = Math.floor((p.y - yMin) / sliceHeight);
    if (s >= sliceCount) s = sliceCount - 1;
    if (s < 0) s = 0;
    sliceCounts[s]++;
    slicePoints[s].push(i);
  }
  let floatingMass = false;
  const anomalyIndices: number[] = [];
  for (let s = 2; s < sliceCount; s++) {
    const ratio = sliceCounts[s] / n;
    const below = sliceCounts[s - 1];
    if (ratio > 0.08 && below === 0) {
      floatingMass = true;
      anomalyIndices.push(...slicePoints[s]);
    }
  }

  // 5. Confidence + status
  const supportScore = centroidSupported ? 1 : 0.2;
  const floatScore = floatingMass ? 0.1 : 1;
  const massScore = Math.min(1, baseRatio / 0.20);
  const confidence = Math.max(0, Math.min(1, 0.45 * supportScore + 0.40 * floatScore + 0.15 * massScore));
  const status: "pass" | "fail" = confidence >= sensitivity && !floatingMass && centroidSupported ? "pass" : "fail";
  const anomaly_detected = floatingMass || !centroidSupported;

  // 6. ZK mock hash — deterministic
  const sorted = [...points].sort((a, b) => a.x - b.x || a.y - b.y || a.z - b.z);
  const payload = JSON.stringify(sorted) + ":" + status;
  const hash = await sha256Hex(payload);

  return {
    status,
    confidence: Math.round(confidence * 1000) / 1000,
    anomaly_detected,
    zk_mock_hash: "0x" + hash,
    metrics: {
      total: n,
      baseCount,
      baseRatio: Math.round(baseRatio * 1000) / 1000,
      centroid: { x: round(cx), y: round(cy), z: round(cz) },
      baseCentroid: { x: round(bcx), z: round(bcz) },
      centroidSupported,
      floatingMass,
      anomalyIndices,
      bounds: {
        min: { x: round(xMin), y: round(yMin), z: round(zMin) },
        max: { x: round(xMax), y: round(yMax), z: round(zMax) },
      },
    },
  };
}

function round(v: number) {
  return Math.round(v * 1000) / 1000;
}

function clamp01(v: number) {
  if (!Number.isFinite(v)) return 0;
  return Math.max(0, Math.min(1, v));
}